src/credentials/providers/credentials.ts

```diff
@@ -0,0 +1,118 @@
+}
+
+/**
+ * Gets the string form of the given `CredentialsProvider`.
+ *
+ * For use in e.g. the statusbar, menus, etc.  Includes:
+ * - credentials source kind
+ * - instance-identifying information (typically the "profile name")
+ *
+ * @param credentials  Value to be formatted.
+ */
+export function asString(credentials: CredentialsId): string {
+    return [credentials.credentialSource, credentials.credentialTypeId].join(
+        CREDENTIALS_PROVIDER_ID_SEPARATOR
+    )
+}
+
+export function fromString(credentials: string): CredentialsId {
+    const separatorPos = credentials.indexOf(CREDENTIALS_PROVIDER_ID_SEPARATOR)
+
+    if (separatorPos === -1) {
+        throw new Error(`Unexpected credentialsId format: ${credentials}`)
+    }
+
+    const credSource = credentials.substring(0, separatorPos)
+    if (!credentialsProviderType.includes(credSource as any)) {
+        throw new Error(`unexpected credential source: ${credSource}`)
+    }
+
+    return {
+        credentialSource: credSource as CredentialsProviderType,
+        credentialTypeId: credentials.substring(separatorPos + 1),
+    }
+}
+
+export function isEqual(idA: CredentialsId, idB: CredentialsId): boolean {
+    return idA.credentialSource === idB.credentialSource && idA.credentialTypeId === idB.credentialTypeId
+}
+
+
+/**
+ * Credentials source type, broadly describes the kind of thing that supplied the credentials.
+ * 
+ * - "profile" is the most common, this is understood as a profile defined in
+ *   `~/.aws/credentials` or something like it, which may be authenticated via
+ *   static values, SSO, MFA, etc.
+ * - "ec2" or "ecs" means "instance credentials" given by the EC2/ECS "metadata service"
+ *   magic endpoint.
+ *   https://docs.aws.amazon.com/sdkref/latest/guide/setting-global-credential_source.html
+ * - "env" means we read the credentials from `AWS_XX` environment variables.
+ *   https://docs.aws.amazon.com/sdkref/latest/guide/environment-variables.html
+ *
+ * Compare the similar concept `telemetry.CredentialSourceId`.
+ */
+export type CredentialsProviderType = typeof credentialsProviderType[number];
+export const credentialsProviderType = ['profile', 'ec2', 'ecs', 'env'] as const;
+
+/**
+ * Lossy map of CredentialsProviderType to telemetry.CredentialSourceId
+ */
+export function credentialsProviderToTelemetryType(o: CredentialsProviderType): telemetry.CredentialSourceId {
+    switch(o) {
+        case 'ec2':
+        case 'ecs':
+            return 'ec2'
+        case 'env':
+            return 'envVars'
+        case 'profile':
+            return 'sharedCredentials'
+        default:
+            return 'other'
+    }
+}
+
+export interface CredentialsProvider {
+    getCredentialsId(): CredentialsId
+    /**
+     * Gets the credential provider type, a coarser form of
+     * telemetry.CredentialSourceId for use in config files and UIs. #1725
+     *
+     * Compare getCredentialsType(). A single _provider_ serves many types of credentials.
```

Comment:
Is this true in practice though? This still feels a bit awkward since currently we are creating a provider instance per unique credential type & ID. Responsibility between the factory and provider seems blurred.

Reply:
Yeah, a "provider" instance only serves a single profile in the current impl. But the comment is referring to providers as a class, i.e. the "SharedCredentialsProvider" can serve SSO, MFA, etc credentials.

> Responsibility between the factory and provider seems blurred.

+1 Hope we can reduce the indirection in these layers.  The role of factory/manager/providers, is not clear to me.